Support flying harvesters

PunkPun opened the pull request:
Adds some of the plumbing needed for docking logic to support aircraft #20168. With this PR openra will support chinook harvesters from Generals

PunkPun committed:
Testcase

abcdefg30 commented on OpenRA.Mods.Common/Activities/DeliverResources.cs:
```diff
@@ -30,6 +32,7 @@ public DeliverResources(Actor self, Actor targetActor = null)
 		{
 			movement = self.Trait<IMove>();
 			harv = self.Trait<Harvester>();
+			aircraft = self.TraitOrDefault<Aircraft>();
```
```suggestion
			aircraft = movement as Aircraft;
```

PunkPun replied:
I've used `move is Aircraft` instead

abcdefg30 commented on OpenRA.Mods.Common/Activities/FindAndDeliverResources.cs:
```diff
@@ -38,7 +38,8 @@ public FindAndDeliverResources(Actor self, Actor deliverActor = null)
 			harv = self.Trait<Harvester>();
 			harvInfo = self.Info.TraitInfo<HarvesterInfo>();
-			mobile = self.Trait<Mobile>();
+			mobile = self.TraitOrDefault<Mobile>();
+			move = self.Trait<IMove>();
```
```suggestion
			move = self.Trait<IMove>();
			mobile = move as Mobile;
```

abcdefg30 commented on OpenRA.Mods.Common/Activities/HarvestResource.cs:
```diff
@@ -32,6 +33,7 @@ public class HarvestResource : Activity
 		public HarvestResource(Actor self, CPos targetCell)
 		{
 			harv = self.Trait<Harvester>();
+			aircraft = self.TraitOrDefault<Aircraft>();
```
```suggestion
			aircraft =move as Aircraft;
```
(And need to re-order this below the definition of `move`.)

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:
```diff
@@ -29,6 +29,9 @@ public class RefineryInfo : TraitInfo, Requires<WithSpriteBodyInfo>, IAcceptReso
 		public readonly CVec DockOffset = CVec.Zero;
 
+		[Desc("If harvester in an aircraft does it need to land to unload resources.")]
+		public readonly bool RequireAircratToLand = true;
```
```suggestion
		public readonly bool RequireAircraftToLand = true;
```

PunkPun replied:
typo fixed

abcdefg30 commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -73,6 +73,9 @@ public class HarvesterInfo : ConditionalTraitInfo, Requires<MobileInfo>
 		public readonly bool QueueFullLoad = false;
 
+		[Desc("If harvester in an aircraft does it need to land to collect resources.")]
+		public readonly bool RequireAircratToLand = true;
```
```suggestion
		public readonly bool RequireAircraftToLand = true;
```

PunkPun commented:
I'll fix it up tomorrow when i get home

PunkPun committed:
Add FindTileInAnnulus

Searches for the closest cell, applying a custom weight. After a tile with the smallest weight within circle is found, search is aborted. This means that if a viable cell was found 2 units away, it won't pick up on cells 3 units away even if their weight is stronger

PunkPun committed:
Support flying harvesters

AspectInteractive commented:
Found a typo in FindAndDeliverResources.cs, `FindTileInCicle` should be `FindTileInCircle`.

Mailaender commented on OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:
```diff
@@ -75,6 +78,7 @@ public class Refinery : INotifyCreated, ITick, IAcceptResources, INotifySold, IN
 		public bool IsDragRequired => info.IsDragRequired;
 		public WVec DragOffset => info.DragOffset;
 		public int DragLength => info.DragLength;
+		public bool RequireAircratToLand => info.RequireAircraftToLand;
```
```suggestion
		public bool RequireAircraftToLand => info.RequireAircraftToLand;
```

Mailaender commented on OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:
```diff
@@ -29,6 +29,9 @@ public class RefineryInfo : TraitInfo, Requires<WithSpriteBodyInfo>, IAcceptReso
 		public readonly CVec DockOffset = CVec.Zero;
 
+		[Desc("If harvester in an aircraft does it need to land to unload resources.")]
+		public readonly bool RequireAircraftToLand = true;
```
This is not even used here and odd for the default mods. Maybe use an existing trait from `Aircraft` or put this onto an `AircraftHarvester`.

PunkPun replied:
We want this to be a choice on a per dock basis, so it needs to stay on `Refinery` until `Dock` class is added

PunkPun commented:
I'll address #16163 here

PunkPun commented:
Closing this PR for now. I have noticed a bunch of issues deeper in the engine that make this PR messier than it needs to be

I will reopen once I have solved those issues or we are closer to finishing the refactor
